contracts/test/integration/backupOracles.sol

```diff
@@ -0,0 +1,293 @@
+        address incorrectPoolToken1 = address(0x4);
+
+        UniswapV3OracleWrapper.OracleConfig
+            memory daiUsdOracleConfig = UniswapV3OracleWrapper.OracleConfig({
+                twapPeriod: twapPeriod,
+                uniswapPool: address(daiUsdcPool),
+                minPoolLiquidity: daiUsdcPool.liquidity(),
+                precision: precision
+            });
+
+        vm.expectRevert(bytes("Incorrect pool for tokens"));
+        daiOracle = new UniswapV3OracleWrapper(
+            address(core),
+            incorrectPoolToken0,
+            incorrectPoolToken1,
+            uniswapMathWrapper,
+            daiUsdOracleConfig
+        );
+    }
+
+    function testAddPoolSupport() public {
+        uint32 highTwapPeriod = 4000;
+
+        UniswapV3OracleWrapper.OracleConfig
+            memory highTwapPeriodConfig = UniswapV3OracleWrapper.OracleConfig({
+                twapPeriod: highTwapPeriod,
+                uniswapPool: address(daiUsdcPool),
+                minPoolLiquidity: daiUsdcPool.liquidity(),
+                precision: precision
+            });
+
+        UniswapV3OracleWrapper highTwapOracle = new UniswapV3OracleWrapper(
+            address(core),
+            dai,
+            usdc,
+            uniswapMathWrapper,
+            highTwapPeriodConfig
+        );
+
+        (, , , , uint16 observationCardinalityNext, , ) = daiUsdcPool.slot0();
+
+        uint32 meanBlockTime = highTwapOracle.meanBlockTime();
+        uint16 expectedCardinalityForMaxTwapPeriod = uint16(
+            highTwapPeriod / meanBlockTime
+        ) + uint16(10);
+        assertEq(
+            observationCardinalityNext,
+            expectedCardinalityForMaxTwapPeriod
+        );
+    }
+
+    /// @notice The below price tests themselves are not strict as price changes
+    /// Exact price is logged and checked manually against spot prices on Uniswap
+
+    /// @dev Each test is designed to test a different path in the price calculation logic.
+    /// There are 4 possible paths that can be taken in the price calculations, depending on
+    /// whether the inversion factors (a) Invert price ratio, b) Invert decimals) are set
+    /// The paths are:
+    /// 1. Invert price ratio: True, Invert decimals: False
+    /// 2. Invert price ratio: True, Invert decimals: True
+    /// 3. Invert price ratio: False, Invert decimals: False
+    /// 4. Invert price ratio: False, Invert decimals: True
+
+    /// token0: usdc
+    /// token1: weth
+    /// inputTokenDecimals (weth): 18
+    /// outputTokenDecimals (usdc): 6
+    /// ----------  Path  --------------
+    /// Invert price ratio: True, Invert decimals: False
+    function testPriceForEthUsdc() public {
+        (Decimal.D256 memory price, bool valid) = ethOracle.read();
+        console.log("eth price:", price.value / precision);
+        assertTrue(valid);
+
+        assertGt(price.value, 2000e18);
+        assertLt(price.value, 5000e18);
```

Comment:
This test is pretty fragile, I would recommend either widening the range to 1k-10k, or just testing that spot price isn't more than 10 percent away from the TWAP price.